src/terminal/parser/OutputStateMachineEngine.cpp

```diff
@@ -1421,92 +1424,152 @@ bool OutputStateMachineEngine::s_ParseColorSpec(const std::wstring_view string,
                                                 DWORD& rgb) noexcept

```

Comment:
I'd love to add some test cases for this method. But I can not find a place to do that? Seems there is no unit tests for this method? Where should I add them?

Reply:
I'd think maybe the best place to do this would be in _OutputEngineTest.cpp_.

You'll want to add an implementation of one of the color table methods (e.g. `SetDefaultForeground`) in the `StatefulDispatch` class, and then save the color parameter that's passed in, and set a flag to indicate that the method has been called.

Then you can add a `TEST_METHOD` looking something like this:

```cpp
auto dispatch = std::make_unique<StatefulDispatch>();
auto pDispatch = dispatch.get();
auto engine = std::make_unique<OutputStateMachineEngine>(std::move(dispatch));
StateMachine mach(std::move(engine));

mach.ProcessString(L"\033]10;rgb:12/34/56\033\\");
VERIFY_IS_TRUE(pDispatch->_setDefaultForeground);
VERIFY_ARE_EQUAL(RGB(0x12, 0x34, 0x56), pDispatch->_defaultForegroundColor);

pDispatch->ClearState();
```

You can add as many test cases to that method as you need, but remember to clear the state after each case.

Reply:
Thanks. But I wonder if I can just test the specific method s_ParseColor since it is used in various place.

Reply:
Ah OK. I guess then you'd need to make the test class a friend of the `OutputStateMachineEngine` class. So at the bottom of the `OutputStateMachineEngine` class, you'd need to add something like:
```cpp
#ifdef UNIT_TESTING
        friend class OutputEngineTest;
#endif
```
Personally I prefer testing through the public interfaces, but I have seen other places in the code that test private methods this way.